File: ui-selenium/src/main/java/org/aeautomation/pages/AccountInformationPage.java
package org.aeautomation.pages;

import org.aeautomation.core.BasePage;
import org.aeautomation.data.Title;
import org.aeautomation.data.UserRegistrationData;
import org.openqa.selenium.By;

/**
 * Page Object representing the Account Information form page (/signup).
 */
public class AccountInformationPage extends BasePage {
    // Header & Section Identifiers
    private final By pageHeader = By.xpath("//b[contains(text(),'Enter Account Information')]");

    // Account Details Locators
    private final By mrTitleRadio = By.xpath("//input[@type='radio' and (@id='id_gender1' or @value='Mr')]");
    private final By mrsTitleRadio = By.xpath("//input[@type='radio' and (@id='id_gender2' or @value='Mrs')]");
    private final By passwordInput = By.xpath("//input[@data-qa='password' or @name='password']");
    private final By daysDropdown = By.xpath("//select[@data-qa='days' or @name='days']");
    private final By monthsDropdown = By.xpath("//select[@data-qa='months' or @name='months']");
    private final By yearsDropdown = By.xpath("//select[@data-qa='years' or @name='years']");
    private final By newsletterCheckbox = By.xpath("//input[@name='newsletter']");
    private final By offersCheckbox = By.xpath("//input[@name='optin']");

    // Address Details Locators
    private final By firstNameInput = By.xpath("//input[@data-qa='first_name' or @name='first_name']");
    private final By lastNameInput = By.xpath("//input[@data-qa='last_name' or @name='last_name']");
    private final By companyInput = By.xpath("//input[@data-qa='company' or @name='company']");
    private final By address1Input = By.xpath("//input[@data-qa='address' or @name='address1']");
    private final By address2Input = By.xpath("//input[@data-qa='address2' or @name='address2']");
    private final By countryDropdown = By.xpath("//select[@data-qa='country' or @name='country']");
    private final By stateInput = By.xpath("//input[@data-qa='state' or @name='state']");
    private final By cityInput = By.xpath("//input[@data-qa='city' or @name='city']");
    private final By zipcodeInput = By.xpath("//input[@data-qa='zipcode' or @name='zipcode']");
    private final By mobileNumberInput = By.xpath("//input[@data-qa='mobile_number' or @name='mobile_number']");
    private final By createAccountButton = By.xpath("//button[@data-qa='create-account']");

    public static final String HEADER_TEXT = "ENTER ACCOUNT INFORMATION";

    /**
     * Retrieves the text from the account information page header.
     *
     * @return Trimmed header text string (e.g., "ENTER ACCOUNT INFORMATION")
     */
    public String getPageHeaderText() {
        return getText(pageHeader);
    }

    /**
     * Selects the title radio button dynamically based on the Title enum provided.
     *
     * @param title Title enum (MR or MRS)
     * @return Current AccountInformationPage instance for method chaining
     */
    public AccountInformationPage selectTitle(Title title) {
        log.info("Selecting title: {}", title);
        switch (title) {
            case MR -> click(mrTitleRadio);
            case MRS -> click(mrsTitleRadio);
        }
        return this;
    }
    /**
     * Enters the password for the new account.
     *
     * @param password Account password
     * @return Current AccountInformationPage instance for method chaining
     */
    public AccountInformationPage enterPassword(String password) {
        type(passwordInput, password);
        return this;
    }

    /**
     * Selects the user's Date of Birth from the day, month, and year dropdowns.
     *
     * @param day   Day value (e.g., "15")
     * @param month Month visible text (e.g., "May")
     * @param year  Year value (e.g., "1990")
     * @return Current AccountInformationPage instance for method chaining
     */
    public AccountInformationPage selectDateOfBirth(String day, String month, String year) {
        selectByVisibleText(daysDropdown, day);
        selectByVisibleText(monthsDropdown, month);
        selectByVisibleText(yearsDropdown, year);
        return this;
    }

    /**
     * Selects both the newsletter and special offers checkboxes.
     *
     * @return Current AccountInformationPage instance for method chaining
     */
    public AccountInformationPage selectNewsletterAndOffers() {
        click(newsletterCheckbox);
        click(offersCheckbox);
        return this;
    }

    /**
     * Fills in all required and optional address details in the registration form.
     *
     * @param data Record containing user profile and address details
     * @return Current AccountInformationPage instance for method chaining
     */
    public AccountInformationPage fillAddressDetails(UserRegistrationData data) {
        type(firstNameInput, data.firstName());
        type(lastNameInput, data.lastName());
        type(companyInput, data.company());
        type(address1Input, data.address1());
        type(address2Input, data.address2());
        selectByVisibleText(countryDropdown, data.country());
        type(stateInput, data.state());
        type(cityInput, data.city());
        type(zipcodeInput, data.zipcode());
        type(mobileNumberInput, data.mobileNumber());
        return this;
    }

    /**
     * Clicks the "Create Account" button to submit the form.
     *
     * @return New instance of AccountCreatedPage
     */
    public AccountCreatedPage clickCreateAccount() {
        click(createAccountButton);
        handleGoogleVignetteAd();
        return new AccountCreatedPage();
    }
}

File: ui-playwright/pages/AccountInformationPage.ts
import { Page, Locator } from '@playwright/test';
import { BasePage } from './BasePage';
import { AccountCreatedPage } from './AccountCreatedPage';

/**
 * Interface representing the required fields to populate account information.
 */
export interface AccountInfoDetails {
    title?: 'Mr' | 'Mrs';
    password: string;
    day: string;
    month: string;
    year: string;
}

/**
 * Interface representing the required fields to populate account address details.
 */
export interface AddressInfoDetails {
    firstName: string;
    lastName: string;
    company: string;
    address1: string;
    address2: string;
    country: string;
    state: string;
    city: string;
    zipcode: string;
    mobileNumber: string;
}

/**
 * Page Object representing the account registration form page ('/signup').
 */
export class AccountInformationPage extends BasePage {

    /** Heading text constant used for verification assertions. */
    public static readonly ENTER_ACCOUNT_INFO_HEADER_TEXT = 'ENTER ACCOUNT INFORMATION';

    /** Form heading locator for account information section (`getByRole`). */
    readonly enterAccountInfoHeading: Locator;

    /** Title radio button locator for 'Mr' (`getByRole`). */
    readonly mrTitleRadio: Locator;

    /** Title radio button locator for 'Mrs' (`getByRole`). */
    readonly mrsTitleRadio: Locator;

    /** Password input field locator (`getByTestId`). */
    readonly passwordInput: Locator;

    /** Day of birth dropdown selector locator (`getByTestId`). */
    readonly daysSelect: Locator;

    /** Month of birth dropdown selector locator (`getByTestId`). */
    readonly monthsSelect: Locator;

    /** Year of birth dropdown selector locator (`getByTestId`). */
    readonly yearsSelect: Locator;

    /** Checkbox locator for newsletter subscription opt-in (`getByTestId`). */
    readonly newsletterCheckbox: Locator;

    /** Checkbox locator for special offers opt-in (`getByTestId`). */
    readonly specialOffersCheckbox: Locator;

    /** First name input field locator (`getByTestId`). */
    readonly firstNameInput: Locator;

    /** Last name input field locator (`getByTestId`). */
    readonly lastNameInput: Locator;

    /** Company name input field locator (`getByTestId`). */
    readonly companyInput: Locator;

    /** Address line 1 input field locator (`getByTestId`). */
    readonly address1Input: Locator;

    /** Address line 2 input field locator (`getByTestId`). */
    readonly address2Input: Locator;

    /** Country dropdown selector locator (`getByTestId`). */
    readonly countrySelect: Locator;

    /** State input field locator (`getByTestId`). */
    readonly stateInput: Locator;

    /** City input field locator (`getByTestId`). */
    readonly cityInput: Locator;

    /** Zipcode input field locator (`getByTestId`). */
    readonly zipcodeInput: Locator;

    /** Mobile number input field locator (`getByTestId`). */
    readonly mobileNumberInput: Locator;

    /** Primary action button locator to submit account registration (`getByRole`). */
    readonly createAccountButton: Locator;

    constructor(page: Page) {
        super(page);

        // Account Information Section
        this.enterAccountInfoHeading = page.getByRole('heading', { name: 'Enter Account Information' });
        this.mrTitleRadio = page.getByRole('radio', { name: 'Mr.' });
        this.mrsTitleRadio = page.getByRole('radio', { name: 'Mrs.' });

        // Personal Details Fields (using data-qa test IDs)
        this.passwordInput = page.getByTestId('password');
        this.daysSelect = page.getByTestId('days');
        this.monthsSelect = page.getByTestId('months');
        this.yearsSelect = page.getByTestId('years');
        this.newsletterCheckbox = page.locator('#newsletter');
        this.specialOffersCheckbox = page.locator('#optin');

        // Address Details Section (using data-qa test IDs)
        this.firstNameInput = page.getByTestId('first_name');
        this.lastNameInput = page.getByTestId('last_name');
        this.companyInput = page.getByTestId('company');
        this.address1Input = page.getByTestId('address');
        this.address2Input = page.getByTestId('address2');
        this.countrySelect = page.getByTestId('country');
        this.stateInput = page.getByTestId('state');
        this.cityInput = page.getByTestId('city');
        this.zipcodeInput = page.getByTestId('zipcode');
        this.mobileNumberInput = page.getByTestId('mobile_number');

        // Submission Action Button
        this.createAccountButton = page.getByRole('button', { name: 'Create Account' });
    }

    /**
     * Selects the title radio button ('Mr' or 'Mrs') based on input.
     * 
     * @param title - Optional title string ('Mr' | 'Mrs') to select
     */
    async selectTitle(title?: 'Mr' | 'Mrs'): Promise<void> {
        if (!title) return;

        if (title === 'Mr') {
            await this.click(this.mrTitleRadio);
        } else if (title === 'Mrs') {
            await this.click(this.mrsTitleRadio);
        }
    }

    /**
     * Fills title, password, date of birth dropdowns, and checks opt-in options.
     * Defaults to 'Mr' if no title is specified in details.
     * 
     * @param details - AccountInfoDetails object containing title, password, and birth date strings
     */
    async fillAccountInformation(details: AccountInfoDetails): Promise<void> {
        await this.selectTitle(details.title || 'Mr');
        await this.type(this.passwordInput, details.password);
        await this.selectByVisibleText(this.daysSelect, details.day);
        await this.selectByVisibleText(this.monthsSelect, details.month);
        await this.selectByVisibleText(this.yearsSelect, details.year);
        await this.click(this.newsletterCheckbox);
        await this.click(this.specialOffersCheckbox);
    }

    /**
     * Fills all personal and billing address form fields.
     * 
     * @param address - AddressInfoDetails object containing address, country, and contact details
     */
    async fillAddressInformation(address: AddressInfoDetails): Promise<void> {
        await this.type(this.firstNameInput, address.firstName);
        await this.type(this.lastNameInput, address.lastName);
        await this.type(this.companyInput, address.company);
        await this.type(this.address1Input, address.address1);
        await this.type(this.address2Input, address.address2);
        await this.selectByVisibleText(this.countrySelect, address.country);
        await this.type(this.stateInput, address.state);
        await this.type(this.cityInput, address.city);
        await this.type(this.zipcodeInput, address.zipcode);
        await this.type(this.mobileNumberInput, address.mobileNumber);
    }

    /**
     * Clicks the 'Create Account' button and navigates to the AccountCreatedPage.
     * 
     * @returns Promise resolving to a new AccountCreatedPage instance
     */
    async clickCreateAccount(): Promise<AccountCreatedPage> {
        await this.click(this.createAccountButton);
        return new AccountCreatedPage(this.page);
    }
}